### PR TITLE
Burning Hands

### PR DESCRIPTION
**Pull request by KazeEspada**
New wizard spell: Burning Hands.
A flamer thrower in the palm of your hand! (uses fireball sprites which can cause confusion).

New wizard spell type for coders:
Special Item, summon a specific type of item.

New Helper Proc:
Get cone(). Get's a list of turfs in a cone!

### CODE REVIEW

File: code/__HELPERS/unsorted.dm [hunk 1]
```diff
@@ -1022,6 +1022,38 @@ B --><-- A
 
 */
 
+/proc/getCone(var/turf/startPoint, var/dir, var/distance)
+	switch(dir)
+		if(1,2,4,8)
```
Comment (RemieRichards) [inline code/__HELPERS/unsorted.dm]: use defines.

File: code/__HELPERS/unsorted.dm [hunk 2]
```diff
@@ -1022,6 +1022,38 @@ B --><-- A
...
+			return getConeDiagonal(startPoint, dir, distance)
+
+/proc/getConeCardinal(var/turf/startPoint, var/dir, var/distance)
+	var/up = turn(dir, 45)
+	var/down = turn(dir, -45)
+	var/list/turfs = new
+	var/turf/uppoint = startPoint
+	var/turf/downpoint = startPoint
+	turfs += startPoint
+	for(var/i=0, i<distance, i++) //Bad number for distance
```
Comment (RemieRichards) [inline code/__HELPERS/unsorted.dm]: ```for(var/i in 0 to distance)``` pls

File: code/__HELPERS/unsorted.dm [hunk 3]
```diff
@@ -1022,6 +1022,38 @@ B --><-- A
...
+		downpoint = get_step(downpoint,down)
+		turfs += block(uppoint, downpoint)
+	return turfs
+
+
+/proc/getConeDiagonal(var/turf/startPoint, var/dir, var/distance)
+	var/up = turn(dir, 45)
+	var/turf/highpoint = startPoint
+	var/list/turfs
+	for(var/i=0, i<distance, i++)
```
Comment (RemieRichards) [inline code/__HELPERS/unsorted.dm]: ```for(var/i in 0 to distance)``` pls

File: code/__HELPERS/unsorted.dm [hunk 4]
```diff
@@ -1022,6 +1022,38 @@ B --><-- A
...
+		highpoint = get_step(highpoint, up)
+	var/across = turn(dir,-90)
+	for(var/i=0, i<distance, i++) // yes, it does need to be two loops.
```
Comment (RemieRichards) [inline code/__HELPERS/unsorted.dm]: ```for(var/i in 0 to distance)``` pls

File: code/game/objects/items/weapons/flamethrower.dm [hunk 5]
```diff
@@ -238,6 +205,56 @@
...
+		if(!T.air)
+			break
+		if(T == previousturf)
+			continue	//so we don't burn the tile we be standin on
+		if(!T.CanAtmosPass(previousturf))
+			break
+		ignite_turf(T,release_amount,gasmixture,multiplier)
+		sleep(delay)
+		previousturf = T
+	return
```
Comment (RemieRichards) [inline code/game/objects/items/weapons/flamethrower.dm]: useless return

File: code/game/objects/items/weapons/flamethrower.dm [hunk 6]
```diff
@@ -238,6 +205,56 @@
...
+			if(removed.gases["plasma"][MOLES])
+				removed.gases["plasma"][MOLES] *= multiplier
+			target.assume_air(removed)
+			//Burn it based on transfered gas
+			target.hotspot_expose((removed.temperature*2) + 380,500)
+	else
+		target.atmos_spawn_air(SPAWN_TOXINS | SPAWN_HEAT, release_amount*multiplier)
+		target.hotspot_expose(1000,500)
+	SSair.add_to_active(target, 0)
+	return
```
Comment (RemieRichards) [inline code/game/objects/items/weapons/flamethrower.dm]: useless return

File: code/game/objects/items/weapons/flamethrower.dm [hunk 7]
```diff
@@ -130,7 +133,7 @@
...
 		return
-	var/dat = text("<TT><B>Flamethrower (<A HREF='?src=\ref[src];light=1'>[lit ? "<font color='red'>Lit</font>" : "Unlit"]</a>)</B><BR>\n Tank Pressure: [ptank.air_contents.return_pressure()]<BR>\nAmount to throw: <A HREF='?src=\ref[src];amount=-100'>-</A> <A HREF='?src=\ref[src];amount=-10'>-</A> <A HREF='?src=\ref[src];amount=-1'>-</A> [throw_amount] <A HREF='?src=\ref[src];amount=1'>+</A> <A HREF='?src=\ref[src];amount=10'>+</A> <A HREF='?src=\ref[src];amount=100'>+</A><BR>\n<A HREF='?src=\ref[src];remove=1'>Remove plasmatank</A> - <A HREF='?src=\ref[src];close=1'>Close</A></TT>")
+	var/dat = text("<TT><B>Flamethrower (<A HREF='?src=\ref[src];light=1'>[lit ? "<font color='red'>Lit</font>" : "Unlit"]</a>)</B><BR>\n Tank Pressure: [ptank.air_contents.return_pressure()]<BR>\nAmount to throw: <A HREF='?src=\ref[src];amount=-1'>-</A> <A HREF='?src=\ref[src];amount=-0.1'>-</A> <A HREF='?src=\ref[src];amount=-0.01'>-</A> [throw_amount] <A HREF='?src=\ref[src];amount=0.01'>+</A> <A HREF='?src=\ref[src];amount=0.1'>+</A> <A HREF='?src=\ref[src];amount=1'>+</A><BR>\n<A HREF='?src=\ref[src];remove=1'>Remove plasmatank</A> - <A HREF='?src=\ref[src];close=1'>Close</A></TT>")
```
Comment (neersighted) [inline code/game/objects/items/weapons/flamethrower.dm]: REEEEEEEEE TGUI PLEASE

File: code/game/objects/items/weapons/flamethrower.dm [hunk 8]
```diff
@@ -238,6 +205,56 @@
 	if(ptank && damage && attack_type == PROJECTILE_ATTACK && prob(15))
 		owner.visible_message("<span class='danger'>[attack_text] hits the fueltank on [owner]'s [src], rupturing it! What a shot!</span>")
 		var/target_turf = get_turf(owner)
-		ignite_turf(target_turf, 100)
+		ignite_turf(target_turf, ptank.air_contents.total_moles(),ptank.air_contents)
```
Comment (neersighted) [inline code/game/objects/items/weapons/flamethrower.dm]: space this please

File: code/game/objects/items/weapons/flamethrower.dm [hunk 9]
```diff
@@ -238,6 +205,56 @@
...
+
+/obj/item/weapon/flamehand/dropped(user)
+	qdel(src)
+
+/obj/item/weapon/flamehand/afterattack(atom/target, mob/user, flag)
+	if(flag)
+		return
+	var/turf/targetTurf = get_turf(target)
+	var/turfList = getline(user, targetTurf)
+	flame_turf(turfList,null,1,5)
```
Comment (neersighted) [inline code/game/objects/items/weapons/flamethrower.dm]: space things out

File: code/game/objects/items/weapons/flamethrower.dm [hunk 10]
```diff
@@ -238,6 +205,56 @@
...
+/atom/proc/flame_turf(turflist, gasmixture, delay, release_amount = 0.75, multiplier = 1)
+	var/turf/previousturf = get_turf(src)
+	for(var/turf/simulated/T in turflist)
+		if(!T.air)
+			break
+		if(T == previousturf)
+			continue	//so we don't burn the tile we be standin on
+		if(!T.CanAtmosPass(previousturf))
+			break
+		ignite_turf(T,release_amount,gasmixture,multiplier)
```
Comment (neersighted) [inline code/game/objects/items/weapons/flamethrower.dm]: spaces

File: code/game/objects/items/weapons/flamethrower.dm [hunk 11]
```diff
@@ -238,6 +205,56 @@
...
+/obj/item/weapon/flamehand/afterattack(atom/target, mob/user, flag)
+	if(flag)
+		return
+	var/turf/targetTurf = get_turf(target)
+	var/turfList = getline(user, targetTurf)
+	flame_turf(turfList,null,1,5)
+
+
+
+/atom/proc/flame_turf(turflist, gasmixture, delay, release_amount = 0.75, multiplier = 1)
```
Comment (neersighted) [inline code/game/objects/items/weapons/flamethrower.dm]: This takes a list of turfs? The signature should be better typed, ie `list/turflist`, `datum/gas_mixture/gasmixture`
And it should be plural, `flame_turfs`

File: code/game/objects/items/weapons/flamethrower.dm [hunk 12]
```diff
@@ -238,6 +205,56 @@
...
+			continue	//so we don't burn the tile we be standin on
+		if(!T.CanAtmosPass(previousturf))
+			break
+		ignite_turf(T,release_amount,gasmixture,multiplier)
+		sleep(delay)
+		previousturf = T
+	return
+
+
+/atom/proc/ignite_turf(turf/simulated/target, release_amount, var/datum/gas_mixture/gasmixture, multiplier)
```
Comment (neersighted) [inline code/game/objects/items/weapons/flamethrower.dm]: Don't put var in proc signatures

File: code/game/objects/items/weapons/flamethrower.dm [hunk 13]
```diff
@@ -238,6 +205,56 @@
...
 		return 1 //It hit the flamethrower, not them
+
+/obj/item/weapon/flamehand
+	name = "burning hand"
+	desc = "Got you to level 3."
+	flags = ABSTRACT
+	w_class = 5
+	icon = 'icons/obj/wizard.dmi'
+	icon_state = "fireball"
+
```
Comment (KorPhaeron) [inline code/game/objects/items/weapons/flamethrower.dm]: What is the itemstate?

**Comment by KorPhaeron** [comment]
Lich+an infinite flamethrower will probably be pretty miserably unfun to fight.

**Comment by Shapsy** [comment]
This doesn't really sound fun to fight against period

File: code/game/objects/items/weapons/flamethrower.dm (hunk 13 above)
Comment (KazeEspada) [inline code/game/objects/items/weapons/flamethrower.dm]: none? its merely a holder object so you can shoot fire out of your hands?

**Comment by Iamgoofball** [comment]
just get a atmos suit/firesuit and some silver sulf patches nerds they're all over the place

File: code/__HELPERS/unsorted.dm (hunk 1 above)
Comment (duncathan) [inline code/__HELPERS/unsorted.dm]: Better yet, kill the switch
`if(dir in cardinal)`

File: code/__HELPERS/unsorted.dm (hunk 3 above)
Comment (duncathan) [inline code/__HELPERS/unsorted.dm]: That'll be off by one, remie; he uses `<` not `<=`
They want either
`for(var/i in 1 to distance)` or
`for(var/i in 0 to distance-1)`
I'd prefer the first one

File: code/__HELPERS/unsorted.dm (hunk 4 above)
Comment (duncathan) [inline code/__HELPERS/unsorted.dm]: See above

File: code/__HELPERS/unsorted.dm (hunk 3 above)
Comment (RemieRichards) [inline code/__HELPERS/unsorted.dm]: butts

File: code/__HELPERS/unsorted.dm (hunk 2 above)
Comment (duncathan) [inline code/__HELPERS/unsorted.dm]: See below
Also what's this comment mean?

File: code/game/objects/items/weapons/flamethrower.dm (hunk 11 above)
Comment (duncathan) [inline code/game/objects/items/weapons/flamethrower.dm]: >giving every atom a proc to flame turfs

**Comment by KorPhaeron** [comment]
@Iamgoofball 

Someone nerfed regular firesuits so they are not fireproof. So you've got a handful of items that actually stop the fire (atmos hardsuit+CE hardsuit+I think maybe 2 atmos firesuits).

Everything you listed also makes you slow, and the lich has 3 other spells to deal with.

So realistically ~5 slow moving people (who don't generally start with weapons), plus maybe a couple of borgs, can fight this guy. The other 30-60 players are shit outta luck.

**Comment by Iamgoofball** [comment]
silver sulf patches dont make you slow

let's just buff fire suits and atmos hardsuits to fire immunity

**Comment by KorPhaeron** [comment]
Good luck putting on those silver sulf patches when he is tasing you or magic missiling you or whatever. It is not just an infinite flamethrower, it is an infinite flamethrower wizard with infinite lives.

**Comment by KazeEspada** [comment]
At the time I was coding this, LSG was still top tier.

**Comment by Iamgoofball** [comment]
we could make liches take triple atmos fire damage

**Comment by KorPhaeron** [comment]
>skeletons take triple burn

???

Also LSG balance was a horrible mistake

**Comment by Iamgoofball** [comment]
make lich skeletons special snowflakes

**Comment by KorPhaeron** [comment]
It was tested and predictably overpowered.

**Comment by kilkun** [comment]
Good. Wizards are supposed to be unrelentingly overpowered. Considering the crew can just wear hardsuits/firesuits to negate it means there's already a decent counter in the game, the crew just has to get used to using it.

**Comment by Shadowlight213** [comment]
@kilkun Not all hardsuits are fireproof. Only the atmos and CE's hardsuit are.

**Comment by MrStonedOne** [comment]
This is 100% balanced


Merge post host.

Nope, not op at all.

Whats that? why am I turning wizard antag on? oh, no reason, no reason at all.

**Comment by kilkun** [comment]
All exosuits are, though.

Goof is right. Buff the other hardsuits and firesuits and the spell will be balanced in the grand scheme of wizardly things.

**Comment by KazeEspada** [comment]
No, only wizards, atmos tech, CE, and Elite Syndie are flameproof.

**Comment by Shapsy** [comment]
even if all hardsuits were fireproof, how exactly is "you must be wearing this to participate in fighting the wizard and not dying instantly" fun

**Comment by KazeEspada** [comment]
I can add a cooldown to the flame to slow it down, I can make it spawn less plasma.

It's based on a real wizard spell.  Also, there are flamethrower fixes thrown into this PR because I was an idiot and fixed those while coding this.

File: code/game/objects/items/weapons/flamethrower.dm (hunk 7 above)
Comment (KazeEspada) [inline code/game/objects/items/weapons/flamethrower.dm]: I dont even know how to write one?

**Comment by Zilenan91** [comment]
Too powerful. Leaves too much residual damage and is too effective on most of the crew, not something the wizard needs.

**Comment by PKPenguin321** [comment]
Honestly? I thought it was kinda fun in the testing round. I was just a crewmember, but narrowly avoiding death by massive fireballs on multiple occasions was a blast.

I can totally see that it's really powerful, but at the same time it was admittedly fun, sooo... :+-1:

**Comment by YakumoChen** [comment]
It leaves plasma everywhere which means if you're not wearing internals at all times you basically die from the residual toxins damage.

Bad enough the wizard was basically invincible and going around setting fires everywhere, but by the time I finally got cloned I immediately died from breathing plasma.

**Comment by KorPhaeron** [comment]
@PKPenguin321 

>but at the same time it was admittedly fun

Probably because it was novel. Imagine seeing this near every wizard round

**Comment by Zilenan91** [comment]
I eventually just blasted him with buckshot on the shuttle, but he had Lich, so if the shuttle hadn't literally been docked he'd have just come right back and killed everyone anyways.

**Comment by Thunder12345** [comment]
Wasn't even a lich

**Comment by Zilenan91** [comment]
he was a lich, his body was a skeleton when he died.

**Comment by KorPhaeron** [comment]
I asked him to try out burning hands for me but failed to explain the spell/tell him to take fire protection, and I felt bad for him, so I made him a normal skeleton (because it wouldn't be much of a test otherwise).

**Comment by kevinz000** [comment]
This is overpowered as all hell. All due to the plasma
If this is like incendiary ammo from shotguns that'd be fine, making hotspots SO IF YOU GOT HIT OR YOU WALK IN YOU GET SET ON FIRE ONCE
But plasma fires means they can set the whole damn station on fire for AGES and fuck up the atmos so badly the station will never be habitable.
Not to mention plasma gives you heavy toxins and shit.

**Comment by Shadowlight213** [comment]
>It's based on a real wizard spell.

Holy shit. You know a real life wizard?

**Comment by The-Albinobigfoot** [comment]
You don't?

**Comment by Iamgoofball** [comment]
Maybe we should fix it to not make plasma.

Just make a hotspot like CLF3 mixing code does.

File: code/modules/spells/spell_types/specialitem.dm [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+	range = -1
+	level_max = 0 //cannot be improved
+	cooldown_min = 100
+	include_user = 1
+
+	action_icon_state = "fireball"
+
+	var/obj/item/specialItem = /obj/item/weapon/bikehorn/rubberducky //Quack
+
+/obj/effect/proc_holder/spell/targeted/specialitem/cast(list/targets,mob/user = usr)
```
Comment (MrStonedOne) [inline code/modules/spells/spell_types/specialitem.dm]: space args out

**Comment by KazeEspada** [comment]
Gonna split this apart and then reopen as seperate pieces.

**Pull request closed without merging**